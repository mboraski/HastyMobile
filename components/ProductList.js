// Third Party Imports
import React, { Component } from 'react';
import { ScrollView } from 'react-native';
import _ from 'lodash';

// Relative Imports
import ProductDetail from './ProductDetail';


class ProductList extends Component {
  state = {
    products: []
  };

  componentWillMount() {
    // TODO: fetch available products once endpoint complete
    this.setState({
      products: [{
        title: 'Redbull',
        thumbnail_image: 'https://facebook.github.io/react/img/logo_og.png',
        price: '$3.49'
      }, {
        title: 'Monster',
        thumbnail_image: 'https://facebook.github.io/react/img/logo_og.png',
        price: '$3.49'
      }, {
        title: 'Gatorade',
        thumbnail_image: 'https://facebook.github.io/react/img/logo_og.png',
        price: '$2.49'
      }, {
        title: 'Water',
        thumbnail_image: 'https://facebook.github.io/react/img/logo_og.png',
        price: '$0.99'
      }]
    });
  }

  renderProducts() {
    // TODO: change key to some other id
    return _.map(this.state.products, product =>
      <ProductDetail
        key={product.title}
        product={product}
        callAddToCart={this.props.callAddToCart}
      />
    );
  }

  render() {
    return (
      <ScrollView>
        {this.renderProducts()}
      </ScrollView>
    );
  }
}

export default ProductList;
