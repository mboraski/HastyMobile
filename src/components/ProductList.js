// Third Party Imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ScrollView, StyleSheet } from 'react-native';
import map from 'lodash.map';

// Relative Imports
import ProductDetail from './ProductDetail';
import { emY } from '../utils/em';

class ProductList extends Component {
    static propTypes = {
        products: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
        searchText: PropTypes.string.isRequired,
        handleAddToCart: PropTypes.func.isRequired,
        handleRequestProduct: PropTypes.func.isRequired,
        productImages: PropTypes.shape({}).isRequired
    };

    filterProducts = () => {
        const { products, searchText } = this.props;
        if (searchText.trim().length) {
            return products.filter(product => {
                return (
                    product.productName
                        .toLowerCase()
                        .indexOf(searchText.toLowerCase()) > -1
                );
            });
        }
        return products;
    };

    renderProducts() {
        const {
            handleAddToCart,
            handleRequestProduct,
            productImages
        } = this.props;
        return map(this.filterProducts(), product => {
            const image = productImages[product.id] || '';
            const quantityTaken = product.quantityTaken;
            const quantityAvailable = product.quantityAvailable;
            const consumed = quantityTaken >= quantityAvailable;
            const inCart = quantityTaken > 0;
            return (
                <ProductDetail
                    key={product.id}
                    consumed={consumed}
                    quantityAvailable={quantityAvailable}
                    quantityTaken={quantityTaken}
                    product={product}
                    inCart={inCart}
                    image={image}
                    handleAddToCart={handleAddToCart}
                    handleRequestProduct={handleRequestProduct}
                />
            );
        });
    }

    render() {
        return (
            <ScrollView contentContainerStyle={styles.container}>
                {this.renderProducts()}
            </ScrollView>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flexWrap: 'wrap',
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'flex-start',
        paddingBottom: emY(6), // to put the last products over checkout button
        margin: 5
    }
});

export default ProductList;
