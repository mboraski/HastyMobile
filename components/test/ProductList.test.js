import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import ProductList from '../ProductList';

const initialState = {};

describe('ProductList', () => {
    const middlewares = [];
    const mockStore = configureStore(middlewares);
    it('renders correctly', () => {
        const wrapper = shallow(<ProductList />, {
            context: { store: mockStore(initialState) }
        });
        const render = wrapper.dive();
        expect(render).toMatchSnapshot();
    });
});
