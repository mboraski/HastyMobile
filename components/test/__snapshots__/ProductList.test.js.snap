// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <ScrollViewMock
    contentContainerStyle={
        Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "paddingLeft": 20,
            "paddingRight": 5,
          }
    }
>
    <ProductDetail
        added={undefined}
        onPress={[Function]}
        product={
            Object {
                "price": "$3.49",
                "productCode": "1",
                "quantity": 1,
                "thumbnail_image": "image",
                "title": "Type 1 code 1",
              }
        }
        style={
            Object {
                "marginBottom": 10,
                "marginRight": 8,
                "width": 354,
              }
        }
    />
</ScrollViewMock>,
  "nodes": Array [
    <ScrollViewMock
      contentContainerStyle={
            Object {
                  "alignItems": "flex-start",
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "paddingLeft": 20,
                  "paddingRight": 5,
                }
      }
>
      <ProductDetail
            added={undefined}
            onPress={[Function]}
            product={
                  Object {
                        "price": "$3.49",
                        "productCode": "1",
                        "quantity": 1,
                        "thumbnail_image": "image",
                        "title": "Type 1 code 1",
                      }
            }
            style={
                  Object {
                        "marginBottom": 10,
                        "marginRight": 8,
                        "width": 354,
                      }
            }
      />
</ScrollViewMock>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ProductList
        cart={
                Object {
                        "products": Object {},
                        "totalProducts": 0,
                      }
        }
        products={
                Array [
                        Object {
                          "price": "$3.49",
                          "productCode": "1",
                          "quantity": 1,
                          "thumbnail_image": "image",
                          "title": "Type 1 code 1",
                        },
                      ]
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ProductList {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "cart": Object {
            "products": Object {},
            "totalProducts": 0,
          },
          "products": Array [
            Object {
              "price": "$3.49",
              "productCode": "1",
              "quantity": 1,
              "thumbnail_image": "image",
              "title": "Type 1 code 1",
            },
          ],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <ScrollViewMock
          contentContainerStyle={
                    Object {
                              "alignItems": "flex-start",
                              "flexDirection": "row",
                              "flexWrap": "wrap",
                              "paddingLeft": 20,
                              "paddingRight": 5,
                            }
          }
>
          <ProductDetail
                    added={undefined}
                    onPress={[Function]}
                    product={
                              Object {
                                        "price": "$3.49",
                                        "productCode": "1",
                                        "quantity": 1,
                                        "thumbnail_image": "image",
                                        "title": "Type 1 code 1",
                                      }
                    }
                    style={
                              Object {
                                        "marginBottom": 10,
                                        "marginRight": 8,
                                        "width": 354,
                                      }
                    }
          />
</ScrollViewMock>,
        "_debugID": 2,
        "_renderedOutput": <ScrollViewMock
          contentContainerStyle={
                    Object {
                              "alignItems": "flex-start",
                              "flexDirection": "row",
                              "flexWrap": "wrap",
                              "paddingLeft": 20,
                              "paddingRight": 5,
                            }
          }
>
          <ProductDetail
                    added={undefined}
                    onPress={[Function]}
                    product={
                              Object {
                                        "price": "$3.49",
                                        "productCode": "1",
                                        "quantity": 1,
                                        "thumbnail_image": "image",
                                        "title": "Type 1 code 1",
                                      }
                    }
                    style={
                              Object {
                                        "marginBottom": 10,
                                        "marginRight": 8,
                                        "width": 354,
                                      }
                    }
          />
</ScrollViewMock>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ProductList
    cart={
        Object {
            "products": Object {},
            "totalProducts": 0,
          }
    }
    products={
        Array [
            Object {
              "price": "$3.49",
              "productCode": "1",
              "quantity": 1,
              "thumbnail_image": "image",
              "title": "Type 1 code 1",
            },
          ]
    }
/>,
}
`;

exports[`ProductList renders correctly with selected item 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <ScrollViewMock
    contentContainerStyle={
        Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "paddingLeft": 20,
            "paddingRight": 5,
          }
    }
>
    <ProductDetail
        added={undefined}
        onPress={[Function]}
        product={
            Object {
                "price": "$3.49",
                "productCode": "1",
                "quantity": 1,
                "thumbnail_image": "image",
                "title": "Type 1 code 1",
              }
        }
        style={
            Object {
                "marginBottom": 10,
                "marginRight": 8,
                "width": 354,
              }
        }
    />
</ScrollViewMock>,
  "nodes": Array [
    <ScrollViewMock
      contentContainerStyle={
            Object {
                  "alignItems": "flex-start",
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "paddingLeft": 20,
                  "paddingRight": 5,
                }
      }
>
      <ProductDetail
            added={undefined}
            onPress={[Function]}
            product={
                  Object {
                        "price": "$3.49",
                        "productCode": "1",
                        "quantity": 1,
                        "thumbnail_image": "image",
                        "title": "Type 1 code 1",
                      }
            }
            style={
                  Object {
                        "marginBottom": 10,
                        "marginRight": 8,
                        "width": 354,
                      }
            }
      />
</ScrollViewMock>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ProductList
        cart={
                Object {
                        "products": Object {
                          "1": Object {
                            "1": Object {
                              "quantity": 1,
                            },
                          },
                        },
                        "totalProducts": 1,
                      }
        }
        products={
                Array [
                        Object {
                          "price": "$3.49",
                          "productCode": "1",
                          "quantity": 1,
                          "thumbnail_image": "image",
                          "title": "Type 1 code 1",
                        },
                      ]
        }
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ProductList {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "cart": Object {
            "products": Object {
              "1": Object {
                "1": Object {
                  "quantity": 1,
                },
              },
            },
            "totalProducts": 1,
          },
          "products": Array [
            Object {
              "price": "$3.49",
              "productCode": "1",
              "quantity": 1,
              "thumbnail_image": "image",
              "title": "Type 1 code 1",
            },
          ],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <ScrollViewMock
          contentContainerStyle={
                    Object {
                              "alignItems": "flex-start",
                              "flexDirection": "row",
                              "flexWrap": "wrap",
                              "paddingLeft": 20,
                              "paddingRight": 5,
                            }
          }
>
          <ProductDetail
                    added={undefined}
                    onPress={[Function]}
                    product={
                              Object {
                                        "price": "$3.49",
                                        "productCode": "1",
                                        "quantity": 1,
                                        "thumbnail_image": "image",
                                        "title": "Type 1 code 1",
                                      }
                    }
                    style={
                              Object {
                                        "marginBottom": 10,
                                        "marginRight": 8,
                                        "width": 354,
                                      }
                    }
          />
</ScrollViewMock>,
        "_debugID": 4,
        "_renderedOutput": <ScrollViewMock
          contentContainerStyle={
                    Object {
                              "alignItems": "flex-start",
                              "flexDirection": "row",
                              "flexWrap": "wrap",
                              "paddingLeft": 20,
                              "paddingRight": 5,
                            }
          }
>
          <ProductDetail
                    added={undefined}
                    onPress={[Function]}
                    product={
                              Object {
                                        "price": "$3.49",
                                        "productCode": "1",
                                        "quantity": 1,
                                        "thumbnail_image": "image",
                                        "title": "Type 1 code 1",
                                      }
                    }
                    style={
                              Object {
                                        "marginBottom": 10,
                                        "marginRight": 8,
                                        "width": 354,
                                      }
                    }
          />
</ScrollViewMock>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ProductList
    cart={
        Object {
            "products": Object {
              "1": Object {
                "1": Object {
                  "quantity": 1,
                },
              },
            },
            "totalProducts": 1,
          }
    }
    products={
        Array [
            Object {
              "price": "$3.49",
              "productCode": "1",
              "quantity": 1,
              "thumbnail_image": "image",
              "title": "Type 1 code 1",
            },
          ]
    }
/>,
}
`;
